sematic/abstract_future.py

```diff
@@ -129,13 +132,26 @@ def __init__(
         self.parent_future: Optional["AbstractFuture"] = None
         self.nested_future: Optional["AbstractFuture"] = None
 
+        try:
+            external_resources: List[ExternalResource] = list(
+                context().private.external_resources  # type: ignore
+            )
+        except NotInSematicFuncError:
+            # the root future is not created inside another func
+            external_resources = []
+
+        # checks in future_context should ensure that this passes
+        # see: https://github.com/sematic-ai/sematic/issues/388
+        assert len(external_resources) <= 1
```

Comment:
This seems misplaced. Coupling a very low-level primitive (`AbstractFuture`) with infrastructure concerns (external resources). I feel this coupling should be held at the resolver level, not all the way down here.

Reply:
keep in mind that you can currently do:

`future.resource_requirements.kubernetes.tolerations` etc.. That's pretty low level infra info, stuffed in a future. I thought keeping symmetry with that was actually desirable rather than otherwise. Plus, I don't think it's actually a bad thing for info about the requirements of a future to be stored on that future, even if those requirements concern infra.

Reply:
As we discussed, I can get convinced by storing `external_resources` on futures, but hear me out and let's think twice.
In the future, I expect there to be a significant amount of logic around external resources (lifecycle, fault tolerance, etc.). Does this warrant a new "middleware" layer that sits between resolvers and futures, is context-aware, and holds the coupling between external resources and futures?
Maybe it is the context itself?

Now I realize that you also resources in `get_resolved_kwargs` so that could break 👆, unless as suggested we remove the resource status check and make futures unaware of resource state.